### PR TITLE
Vulkan: Implement asynchronous frame dumping.

### PR DESCRIPTION
**Pull request by stenzek**
Very brief testing on AMD seems fine. Needs testing on NV/Intel.

---
This change is [Reviewable](https://reviewable.io/reviews/dolphin-emu/dolphin/4435)

### CODE REVIEW

File: Source/Core/VideoBackends/Vulkan/Renderer.cpp [hunk 1]
```diff
@@ -737,22 +737,133 @@ bool Renderer::DrawFrameDump(const EFBRectangle& rc, u32 xfb_addr,
...
+  // The callback here should set fence to null.
+  if (frame.fence != VK_NULL_HANDLE)
+  {
+    g_command_buffer_mgr->WaitForFence(frame.fence);
+    _assert_(frame.fence == VK_NULL_HANDLE);
+  }
+
+  // Copy the now-populated image data to the output file.
+  AVIDump::Frame state = AVIDump::FetchState(frame.ticks);
   FinishFrameData();
```
Comment (stenzek) [inline Source/Core/VideoBackends/Vulkan/Renderer.cpp]: Is this call necessary?

File: Source/Core/VideoBackends/Vulkan/Renderer.cpp [hunk 2]
```diff
@@ -737,22 +737,133 @@ bool Renderer::DrawFrameDump(const EFBRectangle& rc, u32 xfb_addr,
...
+  // Check fence has been signaled.
+  // The callback here should set fence to null.
+  if (frame.fence != VK_NULL_HANDLE)
+  {
+    g_command_buffer_mgr->WaitForFence(frame.fence);
+    _assert_(frame.fence == VK_NULL_HANDLE);
+  }
+
+  // Copy the now-populated image data to the output file.
+  AVIDump::Frame state = AVIDump::FetchState(frame.ticks);
```
Comment (degasus) [inline Source/Core/VideoBackends/Vulkan/Renderer.cpp]: May you not delay the call "FetchState"?

File: Source/Core/VideoBackends/Vulkan/Renderer.cpp [hunk 3]
```diff
@@ -737,22 +737,133 @@ bool Renderer::DrawFrameDump(const EFBRectangle& rc, u32 xfb_addr,
...
+  if (image.pending)
+    WriteFrameDumpImage(index);
+
+  // Ensure the dimensions of the readback texture are sufficient.
+  if (!image.readback_texture || width > image.readback_texture->GetWidth() ||
+      height > image.readback_texture->GetHeight())
+  {
+    // Allocate a new readback texture.
+    // The reset() call is here so that the memory is released before allocating the new texture.
+    image.readback_texture.reset();
```
Comment (degasus) [inline Source/Core/VideoBackends/Vulkan/Renderer.cpp]: Does this invalidate the mapped pointer?

File: Source/Core/VideoBackends/Vulkan/Renderer.h [hunk 4]
```diff
@@ -141,7 +157,19 @@ class Renderer : public ::Renderer
...
-  std::unique_ptr<StagingTexture2D> m_frame_dump_readback_texture;
   VkFramebuffer m_frame_dump_framebuffer = VK_NULL_HANDLE;
+
+  // Readback resources for frame dumping
+  static const size_t FRAME_DUMP_BUFFERED_FRAMES = 2;
+  struct FrameDumpImage
+  {
+    std::unique_ptr<StagingTexture2D> readback_texture;
+    VkFence fence = VK_NULL_HANDLE;
+    u64 ticks = 0;
```
Comment (degasus) [inline Source/Core/VideoBackends/Vulkan/Renderer.h]: AVIDump::Frame state;
instead of ticks

File: Source/Core/VideoBackends/Vulkan/Renderer.cpp [hunk 5]
```diff
@@ -737,22 +737,133 @@ bool Renderer::DrawFrameDump(const EFBRectangle& rc, u32 xfb_addr,
...
   FinishFrameData();
+  DumpFrameData(reinterpret_cast<const u8*>(frame.readback_texture->GetMapPointer()),
+                static_cast<int>(frame.readback_texture->GetWidth()),
+                static_cast<int>(frame.readback_texture->GetHeight()),
+                static_cast<int>(frame.readback_texture->GetRowStride()), state);
+
+  frame.pending = false;
+}
+
+StagingTexture2D* Renderer::PrepareFrameDumpImage(size_t index, u32 width, u32 height, u64 ticks)
```
Comment (degasus) [inline Source/Core/VideoBackends/Vulkan/Renderer.cpp]: This function accesses m_current_frame_dump_image and has an index parameter. This is a bit redundant.

File: Source/Core/VideoBackends/Vulkan/Renderer.cpp [hunk 6]
```diff
@@ -737,22 +737,133 @@ bool Renderer::DrawFrameDump(const EFBRectangle& rc, u32 xfb_addr,
...
+    }
+  }
+
+  // The copy happens immediately after this function returns, so flag this frame as pending.
+  image.fence = g_command_buffer_mgr->GetCurrentCommandBufferFence();
+  image.ticks = ticks;
+  image.pending = true;
+
+  // This frame is queued, so use the next buffer for the next frame.
+  m_current_frame_dump_image = (m_current_frame_dump_image + 1) % FRAME_DUMP_BUFFERED_FRAMES;
```
Comment (degasus) [inline Source/Core/VideoBackends/Vulkan/Renderer.cpp]: Am I correct here?
PrepareFrameDumpImage(m_current_frame_dump_image) may start to encode the image in the buffer m_current_frame_dump_image. Afterwards it returns this buffer and you'll copy by CopyFromImage into this buffer.

I think this line should be moved after WriteFrameDumpImage.

File: Source/Core/VideoBackends/Vulkan/Renderer.cpp (hunk 5 above)
Comment (stenzek) [inline Source/Core/VideoBackends/Vulkan/Renderer.cpp]: Good point.

**Comment by degasus** [comment]
Review status: 0 of 2 files reviewed at latest revision, 2 unresolved discussions.

---

*[Source/Core/VideoBackends/Vulkan/Renderer.cpp, line 808 at r1](https://reviewable.io:443/reviews/dolphin-emu/dolphin/4435#-KW8Hs1ETG6N8eMpKH0F-r1-808:-KW8ZS_b6XKc-0tgoaR0:bcmeoa7) ([raw file](https://github.com/dolphin-emu/dolphin/blob/f3d229f2679463ed839f1a4b260d6079fd043b33/Source/Core/VideoBackends/Vulkan/Renderer.cpp#L808)):*
<details><summary><i>Previously, stenzek (Stenzek) wrote…</i></summary>
> Is this call necessary?
</details>
Hm, I think it's best to move the FinishFrameData() call to the head of PrepareFrameDumpImage().
In theory, if framedumping fails for one frame, the index will be increased anyway. On the next frames, it might happen that oldframe.panding is false, but newframe was pushed to videocommon. So we might miss this call there.

---

*[Source/Core/VideoBackends/Vulkan/Renderer.cpp, line 831 at r1](https://reviewable.io:443/reviews/dolphin-emu/dolphin/4435#-KW8Hs1ETG6N8eMpKH0F-r1-831:-KW8ZZrA16dZ24Mu4mix:b3f52sj) ([raw file](https://github.com/dolphin-emu/dolphin/blob/f3d229f2679463ed839f1a4b260d6079fd043b33/Source/Core/VideoBackends/Vulkan/Renderer.cpp#L831)):*
<details><summary><i>Previously, stenzek (Stenzek) wrote…</i></summary>
> StagingTexture2D handles this by invalidating the caches when a copy is invoked. Could be an issue if that pointer was read somewhere inbetween invalidating and the buffer copy finishing, but we're not doing that in this case.
</details>
Nevermind, I was asking because this pointer was in use by videocommon.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/dolphin-emu/dolphin/4435)*

**Comment by JMC47** [comment]
Tested on my NVIDIA GeForce GTX 1070.  Works extremely well.  Within 1 - 2 FPS of OpenGL's framedump performance.

**Comment by degasus** [comment]
Reviewed 2 of 2 files at r2.
Review status: all files reviewed at latest revision, 3 unresolved discussions.

---

*[Source/Core/VideoBackends/Vulkan/Renderer.cpp, line 782 at r2](https://reviewable.io:443/reviews/dolphin-emu/dolphin/4435#-KW92Crx3YpBYv2Swkml:-KW92Crx3YpBYv2Swkmm:b-cc4az1) ([raw file](https://github.com/dolphin-emu/dolphin/blob/38c3ca5cd4c7a5e4d9d36ff57b9aa88d51fafd3a/Source/Core/VideoBackends/Vulkan/Renderer.cpp#L782)):*
> ```cpp
> }
> 
> void Renderer::OnFrameDumpImageReady(VkFence fence)
> ```

Likely out of scope, but what do you think about calling DumpFrameData here?
We could also add such a callback to the encoding which frees the texture :D

---

*[Source/Core/VideoBackends/Vulkan/Renderer.cpp, line 801 at r2](https://reviewable.io:443/reviews/dolphin-emu/dolphin/4435#-KW91xxupPhI5Ssf9DN1:-KW91xxupPhI5Ssf9DN2:bdr0hrr) ([raw file](https://github.com/dolphin-emu/dolphin/blob/38c3ca5cd4c7a5e4d9d36ff57b9aa88d51fafd3a/Source/Core/VideoBackends/Vulkan/Renderer.cpp#L801)):*
> ```cpp
>   if (frame.fence != VK_NULL_HANDLE)
>   {
>     g_command_buffer_mgr->WaitForFence(frame.fence);
> ```

Is it fine that frame.fence is a race condition here? It might be NULL_HANDLE here. If so, the if() clause might be moved to the WaitForFence instruction

---

*[Source/Core/VideoBackends/Vulkan/Renderer.cpp, line 850 at r2](https://reviewable.io:443/reviews/dolphin-emu/dolphin/4435#-KW91sLWk8NJR805HIcK:-KW91sLWk8NJR805HIcL:b-a21xfr) ([raw file](https://github.com/dolphin-emu/dolphin/blob/38c3ca5cd4c7a5e4d9d36ff57b9aa88d51fafd3a/Source/Core/VideoBackends/Vulkan/Renderer.cpp#L850)):*
> ```cpp
> 
>   // The copy happens immediately after this function returns, so flag this frame as pending.
>   image.fence = g_command_buffer_mgr->GetCurrentCommandBufferFence();
> ```

Is it fine that this fence is before the  CopyFromImage call?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/dolphin-emu/dolphin/4435)*

**Comment by Fallcrest** [comment]
Would this allow PR [4432](https://github.com/dolphin-emu/dolphin/pull/4432) to give Vulkan a performance boost and avoid the performance decrease that it says would happen otherwise?

**Comment by degasus** [comment]
@Fallcrest Yes.

**Comment by stenzek** [comment]
Review status: all files reviewed at latest revision, 3 unresolved discussions.

---

*[Source/Core/VideoBackends/Vulkan/Renderer.cpp, line 782 at r2](https://reviewable.io:443/reviews/dolphin-emu/dolphin/4435#-KW92Crx3YpBYv2Swkml:-KWBCDZHag4TkvNRBFZd:b-ki735y) ([raw file](https://github.com/dolphin-emu/dolphin/blob/38c3ca5cd4c7a5e4d9d36ff57b9aa88d51fafd3a/Source/Core/VideoBackends/Vulkan/Renderer.cpp#L782)):*
<details><summary><i>Previously, degasus (Markus Wick) wrote…</i></summary>
> Likely out of scope, but what do you think about calling DumpFrameData here?
> We could also add such a callback to the encoding which frees the texture :D
</details>
Shouldn't be a problem. Doing it here may mean that DumpFrameData would be called earlier (if another readback happens, for example), but that shouldn't be an issuse?

---

*[Source/Core/VideoBackends/Vulkan/Renderer.cpp, line 801 at r2](https://reviewable.io:443/reviews/dolphin-emu/dolphin/4435#-KW91xxupPhI5Ssf9DN1:-KWBCe1YI2DMaEqbVID4:b-yx452z) ([raw file](https://github.com/dolphin-emu/dolphin/blob/38c3ca5cd4c7a5e4d9d36ff57b9aa88d51fafd3a/Source/Core/VideoBackends/Vulkan/Renderer.cpp#L801)):*
<details><summary><i>Previously, degasus (Markus Wick) wrote…</i></summary>
> Is it fine that frame.fence is a race condition here? It might be NULL_HANDLE here. If so, the if() clause might be moved to the WaitForFence instruction
</details>
WaitForFence invokes the fence signal callbacks, so before the call it'll be non-null, and after WaitForFence it will be null (because OnFrameDumpImageReady is called). All happens on the same thread, so there won't be any races. Perhaps I should document this call better though.

---

*[Source/Core/VideoBackends/Vulkan/Renderer.cpp, line 850 at r2](https://reviewable.io:443/reviews/dolphin-emu/dolphin/4435#-KW91sLWk8NJR805HIcK:-KWBD2-CxzhovAAOZUus:b-eh8r7x) ([raw file](https://github.com/dolphin-emu/dolphin/blob/38c3ca5cd4c7a5e4d9d36ff57b9aa88d51fafd3a/Source/Core/VideoBackends/Vulkan/Renderer.cpp#L850)):*
<details><summary><i>Previously, degasus (Markus Wick) wrote…</i></summary>
> Is it fine that this fence is before the  CopyFromImage call?
</details>
Yup. It's the fence that is signaled when the command buffer containing the CopyFromImage call completes. Though, for clarity, it might be worth refactoring this to include the CopyFromImage call within the Prepare function (probably rename it in this case).

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/dolphin-emu/dolphin/4435)*

**Comment by degasus** [comment]
:lgtm:

---

Review status: all files reviewed at latest revision, 2 unresolved discussions.

---

*[Source/Core/VideoBackends/Vulkan/Renderer.cpp, line 801 at r2](https://reviewable.io:443/reviews/dolphin-emu/dolphin/4435#-KW91xxupPhI5Ssf9DN1:-KWCwlPg4XRK6I6Pgi-T:bgkt28k) ([raw file](https://github.com/dolphin-emu/dolphin/blob/38c3ca5cd4c7a5e4d9d36ff57b9aa88d51fafd3a/Source/Core/VideoBackends/Vulkan/Renderer.cpp#L801)):*
<details><summary><i>Previously, stenzek (Stenzek) wrote…</i></summary>
> WaitForFence invokes the fence signal callbacks, so before the call it'll be non-null, and after WaitForFence it will be null (because OnFrameDumpImageReady is called). All happens on the same thread, so there won't be any races. Perhaps I should document this call better though.
</details>
Oh, sorry. I thought AddFencePointCallback queue on another thread. Nevermind then. So the callback won't be called earlier than WaitForFence?

---

*[Source/Core/VideoBackends/Vulkan/Renderer.cpp, line 850 at r2](https://reviewable.io:443/reviews/dolphin-emu/dolphin/4435#-KW91sLWk8NJR805HIcK:-KWCw7RjWMbUbiSa7lNG:b-ewu731) ([raw file](https://github.com/dolphin-emu/dolphin/blob/38c3ca5cd4c7a5e4d9d36ff57b9aa88d51fafd3a/Source/Core/VideoBackends/Vulkan/Renderer.cpp#L850)):*
<details><summary><i>Previously, stenzek (Stenzek) wrote…</i></summary>
> Yup. It's the fence that is signaled when the command buffer containing the CopyFromImage call completes. Though, for clarity, it might be worth refactoring this to include the CopyFromImage call within the Prepare function (probably rename it in this case).
</details>
Not needed to. I'm just to unfamiliar with Vulkan.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/dolphin-emu/dolphin/4435#-:-KWCx60AFsvsGX3cHLom:bnfp4nl)*

**Pull request merged**